src/nodeViews/createReactNodeViewConstructor.tsx

```diff
@@ -67,6 +76,58 @@ export type ReactNodeViewConstructor = (
   ...args: Parameters<NodeViewConstructor>
 ) => ReactNodeView;
 
+const REACT_PM_KEY = "reactpmkey";
+
+function findNearestRegistryKey(editorView: EditorView, pos: number) {
+  const parentElement = editorView.domAtPos(pos, 0).node as HTMLElement;
+  const nearestElementWithKey = parentElement?.closest(
+    `[data-${REACT_PM_KEY}]`
+  ) as HTMLElement | null;
+
+  return (
+    nearestElementWithKey?.dataset[REACT_PM_KEY] ?? PORTAL_REGISTRY_ROOT_KEY
+  );
+}
+
+/** A queue of registration tasks to be executed */
+const taskQueue: Array<() => void> = [];
+
+/**
+ * Flush the task queue, executing all tasks
+ */
+function flushTaskQueue() {
+  while (taskQueue.length) {
+    // We're iterating over the length of the queue,
+    // so we know that shift will always return a non-null
+    // task.
+    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+    const register = taskQueue.shift()!;
+    register();
+  }
+}
+
+/**
+ * Ensure that this microtask will be executed at the
+ * end of the current event loop.
+ *
+ * If the flush task has not already been queued, queues
+ * it. Then pushes the microtask onto the queue to be flushed.
+ */
+function ensureMicrotask(microtask: () => void) {
+  if (!taskQueue.length) {
+    queueMicrotask(flushTaskQueue);
+  }
+  taskQueue.push(microtask);
+}
+
+/**
+ * Cancels a queued microtask by removing it from the
+ * task queue.
+ */
+function cancelMicrotask(microtask: () => void) {
+  taskQueue.splice(taskQueue.indexOf(microtask), 1);
+}
+
```

Comment:
I'm having second thoughts about this task queue thing and wondering if we're duplicating the work of the React scheduler handling the functional state setter calls.

There's something nice about explicitly deferring the work, but I wish we could control the flush timing better than `queueMicrotask`. Ideally right after the editor view update, inside the same layout effect, I think?

I get that this is hard because the node view constructors are bound to a register function that's defined higher in the component hierarchy. To do what I'm thinking we might need to change `<ProseMirror>` props and handle `useNodeViews` internally. That might actually be awesome, though.

If you're not in a hurry, let's sit with this for a moment. Do you mind if we take a look at what it would look like to push `useNodeViews` down inside `<ProseMirror>` (inner, maybe) such that make we could maybe do some of this in a less complicated way?

Reply:
I see the work you've put in to keep up with the evolving thinking here and very much appreciate you already having done a lot of work to render our thoughts! But if we can keep holding this as a draft for a moment, I think we might be able to do better!

Reply:
I feel ya. It doesn't even feel right to have to rely on the view descriptor tree at all, since ProseMirror already has the node tree fully composed when the node view constructors are being called! I can't say for sure there's a way to do this without relying on the node view descriptor tree, but if there is, I'd love to find it. I've just published this branch as-is as a pre-release version (`0..2.3-0`), so we can keep working on this branch until we feel good about it

Reply:
Thanks!